components/net/image_cache.rs

```diff
@@ -566,9 +676,111 @@ impl ImageCache for ImageCacheImpl {
+            Occupied(mut occupied_entry) => {
+                let task = occupied_entry.get_mut();
+                if task.result.is_some() {
+                    let _ = sender.send(ImageCacheMessage::VectorImageRasterizationCompleted(
+                        pipeline_id,
+                        image_id,
+                        size,
+                    ));
+                } else {
+                    task.listeners.push((pipeline_id, sender));
+                }
+            },
+            Vacant(_) => {
+                warn!("Image rasterization task not found in the cache");
+            },
+        };
+    }
+
+    fn rasterize_vector_image(
+        &self,
+        image_id: PendingImageId,
+        size: DeviceIntSize,
+    ) -> Option<RasterImage> {
+        let mut store = self.store.lock().unwrap();
+        let Some(vector_image) = store.vector_images.get(&image_id).cloned() else {
+            warn!("Unknown image id {image_id:?} requested for rasterization");
+            return None;
+        };
+
+        match store.rasterized_vector_images.entry((image_id, size)) {
+            Occupied(occupied_entry) => {
+                return occupied_entry.get().result.clone();
+            },
+            Vacant(entry) => entry.insert(RasterizationTask {
+                listeners: vec![],
+                result: None,
+            }),
+        };
+
+        let store = self.store.clone();
+        self.thread_pool.spawn(move || {
+            let natural_size = vector_image.svg_tree.size().to_int_size();
+            let requested_size = {
+                let width = size.width.try_into().unwrap_or(0);
+                let height = size.height.try_into().unwrap_or(0);
+                tiny_skia::IntSize::from_wh(width, height).unwrap_or(natural_size)
+            };
+            let transform = tiny_skia::Transform::from_scale(
+                requested_size.width() as f32 / natural_size.width() as f32,
+                requested_size.height() as f32 / natural_size.height() as f32,
+            );
+            let mut pixmap =
+                tiny_skia::Pixmap::new(requested_size.width(), requested_size.height()).unwrap();
+            resvg::render(&vector_image.svg_tree, transform, &mut pixmap.as_mut());
+
+            let frame = ImageFrame {
+                delay: None,
+                bytes: IpcSharedMemory::from_bytes(&pixmap.take()),
+                width: requested_size.width(),
+                height: requested_size.height(),
+            };
+
+            let mut rasterized_image = RasterImage {
+                width: requested_size.width(),
+                height: requested_size.height(),
+                format: PixelFormat::RGBA8,
+                frames: vec![frame],
+                id: None,
+                cors_status: vector_image.cors_status,
+            };
+
+            let mut store = store.lock().unwrap();
+            set_webrender_image_key(&store.compositor_api, &mut rasterized_image);
```

Comment:
It would also be good to move parts of it outside of the critical section, but it involves more code changes, so I have filed https://github.com/servo/servo/issues/37086

Reply:
Thanks!